### DISCUSSION TITLE
Constraint type inference

### DISCUSSION THREAD

**Discussion by rick-de-water** [dotnet/csharplang#6930] [Language Ideas] [answered]
This is something that has been on many people's wish lists for a while, but now with the addition of generic math it has gained a new use case.

Lets say I want to define a generic 3d vector interface:
```csharp
public interface IVector3<T>
{
    T X { get; set; }
    T Y { get; set; }
    T Z { get; set; }
}
```

As well as a `float` implementation for this example:
```csharp
public struct Vector3f : IVector3<float>
{
    public float X { get; set; }
    public float Y { get; set; }
    public float Z { get; set; }
}
```

I then want to define a cross product method that works for any vector, as long as the numeric constraints are met:
```csharp
public static class Vector
{
    public static TVector Cross<TVector, TElement>(TVector v1, TVector v2)
        where TVector : IVector3<TElement>, new()
        where TElement :
            System.Numerics.ISubtractionOperators<TElement, TElement, TElement>,
            System.Numerics.IMultiplyOperators<TElement, TElement, TElement>
    {
        var result = new TVector();
        result.X = v1.Y * v2.Z - v1.Z * v2.Y;
        result.Y = v1.Z * v2.X - v1.X * v2.Z;
        result.Z = v1.X * v2.Y - v1.Y * v2.X;
        return result;
    }
}
```

Perfect! Lets try it out:
```csharp
var v1 = new Vector3f();
var v2 = new Vector3f();
var v3 = Vector.Cross(v1, v2);
```

So close:
> error CS0411: The type arguments for method 'Vector.Cross<TVector, TElement>(TVector, TVector)' cannot be inferred from the usage. Try specifying the type arguments explicitly.

So instead you have to write the following:
```csharp
var v1 = new Vector3f();
var v2 = new Vector3f();
var v3 = Vector.Cross<Vector3f, float>(v1, v2);
```

Which is not something we want to type all the time. We want C# to infer the types for us.

An alternative would be to change the method signature to `Cross(IVector3<TElement> v1, IVector3<TElement> v2)`. This however would lead to boxing, which has significant performance implications.

So I suggest the language team to reconsider constaint type inference. Right now generic math is unable to reach its full potential without it. Vectors and Matrices, Complex numbers and Quaternions, Points, Sizes and Rectangles, these are all multi value math structures that can greatly benefit from this feature.

**Comment by Joe4evr** [comment, marked as answer]
Related/duplicates: #110, #478, #741, #997, https://github.com/dotnet/roslyn/issues/5023 and https://github.com/dotnet/roslyn/issues/15166.

[And here's](https://github.com/dotnet/roslyn/pull/7850#issuecomment-170154270) a comment describing why it's problematic to do.

**Reply by kklocker** [reply]
That comment is 7 years old though, and the other are also old. 
I have been looking for issues related to these ones, but I find it really hard to see what the "latest" development in the area is -- is this decided a dead proposal? 
I, among many others apparently, are really interested in getting better language support for generics.

**Reply by CyrusNajmabadi** [reply]
>  is this decided a dead proposal?

No suitable proposal has been made so far :(

 If one were made, it would have to address the issues brought up in the past as to why this is very challenging to do :)

**Reply by kklocker** [reply]
Understandable :) It is a bummer still...

I would for instance love beeing able to do something like this: (example usage)
```csharp
interface IReader<TModel> where TModel : ITypedModel<TId>
{
  TModel Get(TId id); // infer TId from constraint
}
```
instead of

```csharp
interface IReader<TModel, TId> where TModel : ITypedModel<TId>
{
  TModel Get(TId id); 
}
```

**Reply by thomhurst** [reply]
Could it not be opt-in functionality based on a project property or something? If people explicitly opt-in them that's on them to be aware about breaking changes .

Like how we have 
`<Nullable>enabled</Nullable>` Or `<ImplicitUsings>enabled</ImplicitUsings>`

We could have something like:

`<EnhancedTypeInferance>enabled</EnhancedTypeInferance>`

**Reply by 333fred** [reply]
This would not meet the criteria for a flag. Flags create language dialects, and we explicitly do not like that. Nullable rose to that standard for being one of the most pervasive problems in computing. Usings is not a compiler flag: it creates a file with a bunch of `global usings` in it, just like you can do. A better hope would be for the breaking change feature to allow us to do this, but detecting breaks from this and fixing them up may be extremely difficult to do. I honestly don't have a ton of hope that we'd implement this unless @agocke manages to come up with some clever idea.

**Reply by agocke** [reply]
Well, not sure if it's clever or crazy, but I had an idea.

The constraint on the language side is that we're not allowed to break overload resolution for all C# customers.

But there's nothing saying we can't add a feature to let users change overload resolution _for their own overloads_. That's basically the idea behind the proposed `OverloadResolutionPriority` attribute.

What if we add a new attribute, `InferConstraintsAttribute` or something similar, and let people apply it to their own methods. When they do, the C# compiler will infer generic constraints for that method. If the inference succeeds and the result is "better" than other overloads, that one gets picked.

**Reply by davidroth** [reply]
Could somebody answer the comment from here https://github.com/dotnet/roslyn/issues/5023#issuecomment-987672740?

He questioned why this was ruled out as a breaking change as it would be possible to fallback to the old behaviour of there are ambiguous overloads.

Quote:

> Why was this ruled out as a breaking change ?
> it appears the only case where this would be a breaking change is if there is ambiguity, so why not fallback to old behavior of needing to explicitly define all type arguments when there is ambiguity ?
> 
> so for this case:
> 
> void M(object) {}
> void M<T, U>(T t) where T : IEnumerable<U> {}
> 
> M("foo")
> 1. determine if there are overloads for M with ambiguous constraints
> 2. if no: assign the type arguments based on input
> 3. if yes: check for a flag (e.g. PreferMostStrictTypeArgumentOverload)
> 4. if flag is set: use the M<T,U> overload
> 5. if flag is not set (default): use the M(object) overload

**Reply by 333fred** [reply]
@davidroth again, a global flag is not an acceptable solution.

@agocke you're right, I hate it. 🤣

**Reply by davidroth** [reply]
@333fred What about just fallbacking and not introducing a flag? 

So for this case:

```
void M(object) {}
void M<T, U>(T t) where T : IEnumerable {}
```

`M("foo")`

determine if there are overloads for M with ambiguous constraints

1. if no: assign the type arguments based on input
2. if yes: use the M(object) overload


No flag, it would only work for new apis where there is no overload with (object)

**Reply by agocke** [reply]
> determine if there are overloads for M with ambiguous constraints

There's no ambiguity in the original examples. `object` is better because `T` fails type inference. Resolution succeeds against `object` with no other alternatives.

**Reply by 333fred** [reply]
> But there's nothing saying we can't add a feature to let users change overload resolution for their own overloads. That's basically the idea behind the proposed OverloadResolutionPriority attribute.

The key difference to me is that OverloadResolutionPriority is a "hopefully you never need it" attribute, whereas this would be a "why aren't you putting this on all your methods" attribute. Probably more nuisance than useful overall.

**Reply by agocke** [reply]
So the problem with `ImprovedInferenceAttribute` is that it's _too good_ and the other methods might get jealous? 😉

Just saying, we could also allow targeting the assembly...

**Reply by thomhurst** [reply]
Suggestion here for an `implicit` keyword:
https://github.com/dotnet/roslyn/pull/7850#issuecomment-2196709313

**Comment by thomhurst** [comment]
I'm building a library with some complicated constraints. Due to this lack of automatic inference, some of the things I want to do have become undo-able because I can't expect the caller to have to put in my complicated generic arguments. See #8205

**Comment by neuecc** [comment]
With the introduction of `static abstract members` and, more recently, allowing `ref struct`, the importance of generics has increased compared to the past.
It's strange that opinions from 10 years ago continue to reject this.
Is there no room for reconsideration?

**Comment by jnm2** [comment]
Partial type inference is being considered (https://github.com/dotnet/csharplang/discussions/8967), which would allow syntax something like:

```cs
var v3 = Vector.Cross<_, _>(v1, v2);
```

This avoids the compatibility issue since the `<_, _>` prevents the compiler from calling a method with a different generic arity.

**Comment by FelixCCWork** [comment]
I really like the idea of an implicit opt-in mentioned here https://github.com/dotnet/roslyn/pull/7850#issuecomment-2196709313

> Couldn't this simply be solved by specifying that U should be inferred from the where clause if not given explicitly? For example using a reserved word:
> 
> ```
> void M(object) {}
> void M<T, U>(T t) where T : IEnumerable<U> where U : implicit {} // U is inferred from where clause
> // alternatively: 
> void M<T, implicit U>(T t) where T : IEnumerable<U> // U is inferred from where clause
> ```
> 
> This wouldn't break anything, would it? Just enable new functionality on new code.

From a library author perspective, this feature should be as easy to use for consumers as possible. Having to use something like `<_, _>`  seems not a huge improvement. In my opinion, it's worth for libraries having to put in some more work like using the implicit approach mentioned above.
